spi/src/main/java/com/amazon/opendistroforelasticsearch/jobscheduler/spi/utils/LockService.java

```diff
@@ -0,0 +1,234 @@
+        this.client = client;
+        this.clusterService = clusterService;
+    }
+
+    private String lockMapping() {
+        try {
+            InputStream in = LockService.class.getResourceAsStream("opendistro_job_scheduler_lock.json");
+            StringBuilder stringBuilder = new StringBuilder();
+            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+            for (String line; (line = bufferedReader.readLine()) != null; ) {
+                stringBuilder.append(line);
+            }
+            return stringBuilder.toString();
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Lock Mapping cannot be read correctly.");
+        }
+    }
+
+    public boolean lockIndexExist() {
+        return clusterService.state().routingTable().hasIndex(LOCK_INDEX_NAME);
+    }
+
+    @VisibleForTesting
+    boolean createLockIndex() {
+        final boolean exists = clusterService.state().routingTable().hasIndex(LOCK_INDEX_NAME);
+        if (exists) {
+            return true;
+        }
+
+        try {
+            final CreateIndexRequest request = new CreateIndexRequest(LOCK_INDEX_NAME)
+                .mapping(MAPPING_TYPE, lockMapping(), XContentType.JSON);
+            return client.admin().indices().create(request).actionGet().isAcknowledged();
+        } catch (ResourceAlreadyExistsException e) {
+            return true;
+        }
+    }
+
+    public LockModel acquireLock(final String jobIndexName, final String jobId, final long lockDurationSecond) {
+        if (!lockIndexExist()) {
+            createLockIndex();
+        }
+
+        try {
+            LockModel existingLock = findLock(LockModel.generateLockId(jobIndexName, jobId));
+            if (existingLock != null) {
+                if (isLockReleasedOrExpired(existingLock)) {
+                    // Lock is expired. Attempt to acquire lock.
+                    logger.debug("lock is released or expired: " + existingLock);
+                    LockModel updateLock = new LockModel(existingLock, getNow(), lockDurationSecond, false);
+                    return updateLock(updateLock);
+                } else {
+                    logger.debug("Lock is NOT released or expired. " + existingLock);
+                    // Lock is still not expired. Return null as we cannot acquire lock.
+                    return null;
+                }
+            } else {
+                // There is no lock object and it is first time. Create new lock.
+                LockModel tempLock = new LockModel(jobIndexName, jobId, getNow(), lockDurationSecond, false);
+                logger.debug("Lock does not exist. Creating new lock" + tempLock);
+                return createLock(tempLock);
+            }
+        } catch (VersionConflictEngineException e) {
+            logger.debug("could not acquire lock {}", e.getMessage());
```

Comment:
we should log job index and ID in the message, or do we have them in the exception message?

Reply:
The exception message contains the `lockId` aka DocumentId for the lock index. From this we can derive JobIndex and JobId.